test/parallel_api/algorithm/alg.modifying.operations/shift_left_right.pass.cpp

```diff
@@ -0,0 +1,178 @@
+        //Otherwise, returns last.
+        It __last = std::next(first, m);
+        auto res_exp = (n > 0 && n < m ? std::next(first, m - n) : (n > 0 ? first : __last));
+
+        EXPECT_TRUE(res_exp == res, "wrong return value of shift_left");
+
+        if(res != first && res != __last)
+        {
+            EXPECT_EQ_N(first, std::next(first_exp, + n), m - n, "wrong effect of shift_left");
+            //restore unput data
+            std::copy_n(first_exp, m, first);
+        }
+    }
+};
+
+struct shift_right_algo
+{
+    template <typename Policy, typename It, typename Size>
+    typename ::std::enable_if<TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value,
+                            It>::type
+    operator()(Policy&& exec, It first, It last, Size n)
+    {
+        return oneapi::dpl::shift_right(::std::forward<Policy>(exec), first, last, n);
+    }
+    //skip the test for non-bidirectional iterator (forward iterator, etc)
+    template <typename Policy, typename It, typename Size>
+    typename ::std::enable_if<!TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value,
+                            It>::type
+    operator()(Policy&& exec, It first, It last, Size n)
+    {
+        return first;
+    }
+
+    template <typename It, typename Size>
+    typename ::std::enable_if<TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value,
+                            void>::type
+    check(It res, It first, Size m, It first_exp, Size n)
+    {
+        //if (n > 0 && n < m), returns first + n. Otherwise, if n  > 0, returns last.
+        //Otherwise, returns firts.
+        It __last = std::next(first, m);
+        auto res_exp = (n > 0 && n < m ? std::next(first, n) : (n > 0 ? __last : first));
+
+        EXPECT_TRUE(res_exp == res, "wrong return value of shift_right");
+
+        if(res != first && res != __last)
+        {
+            EXPECT_EQ_N(::std::next(first, n), first_exp, m - n, "wrong effect of shift_right");
+            //restore unput data
+            std::copy_n(first_exp, m, first);
+        }
+    }
+    //skip the check for non-bidirectional iterator (forward iterator, etc)
+    template <typename It, typename Size>
+    typename ::std::enable_if<!TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value,
+                            void>::type
+    check(It res, It first, Size m, It first_exp, Size n)
+    {
+    }
+};
+
+int
+main()
+{
+    const ::std::size_t N = 100000;
+    for (long m = 0; m < N; m = m < 16 ? m + 1 : long(3.1415 * m))
+        for (long n = 0; n < N; n = n < 16 ? n + 1 : long(3.1415 * n))
+    {
+        //std::cout << "m: " << m << " n: " << n << std::endl;
+#ifdef _PSTL_TEST_SHIFT_LEFT
+        test_shift_by_type<int32_t>(m, n, shift_left_algo{}); 
+#endif
+#ifdef _PSTL_TEST_SHIFT_RIGHT
+       test_shift_by_type<int32_t>(m, n, shift_right_algo{});
```

Comment:
I think for coverage it's better if we use here a different type, e.g. `double`

Reply:
no, I don't think so. We have not compute (floating arithmetic) in the test and it is not perfomance test - so doesn't matter  - "moving"  4 bytes or 8.